### PR TITLE
Activated task functionality

### PR DESCRIPTION
**Pull request by mminella**
This commit enables task functionality for Spring Cloud Data Flow on
CloudFoundry.  CloudFoundry 1.7.12+ is required for task support.

Resolves spring-cloud/spring-cloud-dataflow-server-cloudfoundry#43

Depends on spring-cloud/spring-cloud-deployer-cloudfoundry#63

### CODE REVIEW

File: spring-cloud-dataflow-server-cloudfoundry/src/main/resources/dataflow-server.yml [hunk 1]
```diff
@@ -39,7 +39,16 @@ spring:
       uri: http://localhost:8888
     dataflow:
       features:
-        tasksEnabled: false
+        tasksEnabled: true
+    deployer:
+      cloudfoundry:
+        taskTimeout: 360
+#  datasource:
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-cloudfoundry/src/main/resources/dataflow-server.yml]: Probably want to remove these credentials

**Comment by sabbyanandan** [comment]
I tried this with spring-cloud/spring-cloud-deployer-cloudfoundry#63 on PWS and I was able to successfully create and launch Tasks. One of the Task variant included multiple batch-jobs in it, too.

**Comment by trisberg** [comment]
tried this on "ivory" and I'm getting:

```
dataflow:>task launch t3
Command failed org.springframework.cloud.dataflow.rest.client.DataFlowClientException: Source emitted more than one item
```

**Comment by ericbottard** [comment]
I tried this on api.a1.cf-app.com, (built deployer skipping tests for now) and was able to launch timestamp task. The dumb question though is, how can I know that all went well?
Updated my cf client to the latest edge, but I don't see any command that could help me seeing task based processes.

**Comment by trisberg** [comment]
Yes, how can you view the logs of the completed tasks to verify that it succeeded?

**Comment by sabbyanandan** [comment]
With the [v3 cf-cli](https://github.com/cloudfoundry/v3-cli-plugin) plugin installed, as soon as I launch the task, I could tail the logs via: `cf v3-logs <APP_NAME>`. You'd be able to tail only at the launch event, because the container gets created only then.

I didn't try `cf v3-logs <APP_NAME> --recent` - this might work. 

O'wise, you can verify executions via `task execution list` command or by directly querying the `task_execution` table.

**Comment by trisberg** [comment]
Using `--recent` with `v3-logs` is not an option, so there seems to be no way of launcing a task and looking at the logs. Having to tail the app logs before launching the task seems to me to be the wrong approach.

**Comment by ericbottard** [comment]
This is what I get, starting from a clean slate:

```
dataflow:>app import http://bit.ly/task-applications-maven
Successfully registered applications: [task.spark-yarn, task.spark-cluster, task.spark-client, task.timestamp]
dataflow:>task create mytask --definition "timestamp --format=\"yyyy\""
Created new task 'mytask'
dataflow:>task launch mytask
Launched task 'mytask'
dataflow:>task list
╔═════════╤═════════════════════════╤═══════════╗
║Task Name│     Task Definition     │Task Status║
╠═════════╪═════════════════════════╪═══════════╣
║mytask   │timestamp --format="yyyy"│complete   ║
╚═════════╧═════════════════════════╧═══════════╝

dataflow:>task execution list
╔═════════╤══╤══════════╤════════╤═════════╗
║Task Name│ID│Start Time│End Time│Exit Code║
╚═════════╧══╧══════════╧════════╧═════════╝

dataflow:>task status --name mytask
Feature Not Available
```

I don't think the last two oddities are CF specific though.
Running `task list` does trigger CF API chat, as expected

**Comment by trisberg** [comment]
I get the same message from `task status` on k8s, the `task execution list` works though.

**Comment by sabbyanandan** [comment]
@ericbottard: I ran into a similar issue, but then I realized the `timestamp` task was using `h2` database instead of `mysql`. You can see that in the log when the "autoconfiguration" kicks in. 

Make sure, you have something similar configured at the server level. 
> cf set-env sabby-dataflow-server SPRING_CLOUD_DEPLOYER_CLOUDFOUNDRY_SERVICES redis,rabbit,mysql

After the server restage/restart, I was able to see the records in `mysql` and then in the `task execution list`, too.

**Comment by ericbottard** [comment]
@sabbyanandan Good catch, even though it doesn't seem to work for me, even after adding `my_mysql` to the list of services. BTW, should we use a separate list of services to bind for streams and tasks? cc @mminella @markfisher 

I could not access the logs of my task. The v3 plugin does not seem to work very well for me (`v3-apps` fails, had to use `v3apps`, no dash. Then `v3-logs` complains the app (that it just listed) does not exist...)

**Comment by sabbyanandan** [comment]
I think we need to document the DB requirement for tasks explicitly. Also, we would have to include Task docs in the [reference guide](http://docs.spring.io/spring-cloud-dataflow-server-cloudfoundry/docs/1.0.0.M4/reference/htmlsingle/#overview) - only streams included now. 

Good model to follow is what we have done in the k8s [reference guide](http://docs.spring.io/spring-cloud-dataflow-server-kubernetes/docs/1.0.0.RELEASE/reference/htmlsingle/) - we need the same includes.

**Comment by markfisher** [comment]
yes, we need to make a distinction for common services at the stream/task level. For all stream apps, we need to bind to the common service that is used by the binders (rabbit or kafka), and for all task apps we need to bind to the DB that is used as a common service for the task execution state. Individual (non-common) services (such as the DB used by a JDBC-based stream app) should still be bindable via a config option at the level of the individual app.

**Pull request merged**